ouroboros-consensus/src/Ouroboros/Consensus/Util/MonadSTM/StrictMVar.hs

```diff
@@ -0,0 +1,151 @@
+import           Control.Monad.Class.MonadSTM.Strict (checkInvariant)
+import           GHC.Stack
+
+{-------------------------------------------------------------------------------
+  Strict MVar
+-------------------------------------------------------------------------------}
+
+-- | Strict MVar (modelled using a lazy 'Lazy.TMVar' under the hood)
+--
+-- The 'StrictMVar' API is slightly stronger than the usual 'MVar' one, as we
+-- offer a primitive to read the value of the MVar even if it is empty (in which
+-- case we will return the oldest known stale one). See 'readMVarSTM'.
+--
+-- There is a weaker invariant for a 'StrictMVar' than for a 'StrictTVar':
+-- although all functions that modify the 'StrictMVar' check the invariant, we
+-- do /not/ guarantee that the value inside the 'StrictMVar' always satisfies
+-- the invariant. Instead, we /do/ guarantee that if the 'StrictMVar' is updated
+-- with a value that does not satisfy the invariant, an exception is thrown. The
+-- reason for this weaker guarantee is that leaving an 'MVar' empty can lead to
+-- very hard to debug "blocked indefinitely" problems.
+--
+-- This is also the reason we do not offer support for an invariant in
+-- 'StrictTMVar': if we throw an exception from an STM transaction, the STM
+-- transaction is not executed, and so we would not even be able to provide the
+-- weaker guarantee that we provide for 'StrictMVar'.
+data StrictMVar m a = StrictMVar
+  { invariant :: !(a -> Maybe String)
+    -- ^ Invariant checked whenever updating the 'StrictMVar'.
+  , tmvar     :: !(Lazy.LazyTMVar m a)
+    -- ^ The main TMVar supporting this 'StrictMVar'
+  , tvar      :: !(Lazy.LazyTVar m a)
+    -- ^ TVar for supporting 'readOldValue'
+    --
+    -- This TVar is always kept up to date with the 'LazyTMVar', but holds on
+    -- the old value of the 'LazyTMVar' when it is empty. This is very useful
+    -- to support single writer/many reader scenarios.
+    --
+    -- NOTE: We should always update the 'tmvar' before the 'tvar' so that if
+    -- the update to the 'tmvar' fails, the 'tvar is left unchanged.
+  }
+
+newMVar :: MonadSTM m => a -> m (StrictMVar m a)
+newMVar = newMVarWithInvariant (const Nothing)
+
+newMVarWithInvariant :: (MonadSTM m, HasCallStack)
+                     => (a -> Maybe String)  -- ^ Invariant (expect 'Nothing')
+                     -> a
+                     -> m (StrictMVar m a)
+newMVarWithInvariant invariant !a =
+    checkInvariant (invariant a) $
+    StrictMVar invariant <$> Lazy.newTMVarM a <*> Lazy.newTVarM a
+
+newEmptyMVar :: MonadSTM m => a -> m (StrictMVar m a)
+newEmptyMVar = newEmptyMVarWithInvariant (const Nothing)
+
+newEmptyMVarWithInvariant :: MonadSTM m
+                          => (a -> Maybe String)  -- ^ Invariant (expect 'Nothing')
+                          -> a                    -- ^ The initial stale value
+                          -> m (StrictMVar m a)
+newEmptyMVarWithInvariant invariant stale =
```

Comment:
Also kind of subtle that we're not strict in `a` to allow for `error`. Maybe this pattern should be  documented?